pkg/auth/authz_handler.go

```diff
@@ -0,0 +1,139 @@
+package auth
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"regexp"
+	"strings"
+
+	"github.com/openshift/assisted-service/internal/common"
+	"github.com/openshift/assisted-service/pkg/ocm"
+	"github.com/openshift/assisted-service/restapi"
+	"github.com/patrickmn/go-cache"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	bareMetalClusterResource string = "BareMetalCluster"
+	amsActionCreate          string = "create"
+	clustersPathRegex        string = "clusters/.+"
+	capabilityName           string = "bare_metal_installer_admin"
+	capabilityType           string = "Account"
+)
+
+// CreateAuthorizer returns Authorizer if auth is enabled
+func (a *AuthHandler) CreateAuthorizer() func(*http.Request) error {
+	if !a.EnableAuth {
+		return func(*http.Request) error {
+			return nil
+		}
+	}
+
+	return a.Authorizer
+}
+
+// Authorizer is used to authorize a request after the Auth function was called using the "Auth*" functions
+// and the principal was stored in the context in the "AuthKey" context value.
+func (a *AuthHandler) Authorizer(request *http.Request) error {
+	payload := PayloadFromContext(request.Context())
+	if payload == nil || payload.Username == "" {
+		// auth is disabled
+		return nil
+	}
+	username := payload.Username
+	payloadFromCache, found := a.client.Cache.Get(username)
+	if found {
+		// Update payload with values from cache
+		payloadFromCache := payloadFromCache.(*ocm.AuthPayload)
+		payload.IsUser = payloadFromCache.IsUser
+		payload.IsAdmin = payloadFromCache.IsAdmin
+	} else {
+		allowed, err := a.allowedToUseAssistedInstaller(username)
+		if err != nil {
+			logrus.Errorf("Failed to authorize user: %v", err)
+		}
+		payload.IsUser = allowed
+		a.client.Cache.Set(username, payload, cache.DefaultExpiration)
+	}
+
+	if payload.IsUser {
+		if payload.IsAdmin {
+			// All endpoints are allowed for admins.
+			return nil
+		}
+
+		// If endpoint contains a cluster ID, ensures it's owned by the user.
+		if clusterID := a.getClusterIDFromPath(request.URL.Path); clusterID != "" {
```

Comment:
I don't think this should be done in the authorizer.
This filtering should be done in the endpoint handler itself.
For example, in List Cluster operation, you have to do it in the handler

Reply:
This check is actually to avoid manipulating a cluster not owned by the user, the filterring is done in the db query:
https://github.com/openshift/assisted-service/pull/73/files#diff-97c2fe56dacc4fee9ab38e4dbd281ab8R1073

Reply:
DB should not be accessed from the middleware

Reply:
ok, moved to handlers then.